### PR TITLE
CompatOverride API

### LINKED ISSUE #7322: Expose CompatOverrides via API
https://bugzilla.mozilla.org/show_bug.cgi?id=1402064#c16 highlighted this functionality is provided (but not documented) in the v1.5 API and is still needed by Firefox Addon Manager so we need to expose it in the v3 API too.

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #7322

### CODE REVIEW

File: src/olympia/addons/views.py [hunk 1]
```diff
@@ -822,3 +823,16 @@ class ReplacementAddonView(ListAPIView):
...
+
+class CompatOverrideView(ListAPIView):
+    queryset = CompatOverride.objects.all()
+    serializer_class = CompatOverrideSerializer
+
+    def filter_queryset(self, queryset):
+        # Use the same Filter we use for AddonSearchView for consistency.
+        guid_filter = AddonGuidQueryParam(self.request)
+        guids = guid_filter.get_value()
+        print guids
```
Comment (diox) [inline src/olympia/addons/views.py]: extra print

File: src/olympia/addons/tests/test_views.py [hunk 2]
```diff
@@ -3292,3 +3292,57 @@ def test_basic(self):
...
+        self.addon = addon_factory(guid='extrabad@thing')
+        self.override_addon = CompatOverride.objects.create(
+            name='override with addon', guid=self.addon.guid, addon=self.addon)
+        CompatOverrideRange.objects.create(
+            compat=self.override_addon, app=amo.FIREFOX.id)
+        self.override_without = CompatOverride.objects.create(
+            name='override no addon', guid='bad@thing')
+        CompatOverrideRange.objects.create(
+            compat=self.override_without, app=amo.FIREFOX.id)
+
```
Comment (diox) [inline src/olympia/addons/tests/test_views.py]: Add a test with no guid passed, it should return a 400

File: docs/topics/api/addons.rst [hunk 3]
```diff
@@ -490,3 +490,35 @@ This endpoint returns a list of suggested replacements for legacy add-ons that a
...
     :>json string results[].replacement[]: An array of guids for the replacements add-ons.  If there is a direct replacement this will be a list of one add-on guid.  The list can be empty if all the replacement add-ons are invalid (e.g. not publicly available on AMO).  The list will also be empty if the replacement is to a url that is not an addon or collection.
+
+
+----------------
+Compat Overrides
+----------------
+
+.. _addon-compat-overrides:
+
+This endpoint allows compatability overrides specified by AMO admins to be searched.
```
Comment (diox) [inline docs/topics/api/addons.rst]: Typo "compatability", also repeated a couple times below.

File: docs/topics/api/addons.rst [hunk 4]
```diff
@@ -490,3 +490,35 @@ This endpoint returns a list of suggested replacements for legacy add-ons that a
...
+Compatabilty overrides are used within Firefox (and other toolkit applications e.g. Thunderbird) to change compatability of installed add-ons where they have stopped working correctly in new release of Firefox, etc.
+
+.. http:get:: /api/v3/addons/compat-override/
```
Comment (diox) [inline docs/topics/api/addons.rst]: The title and anchor and description are using plural, but this is singular. Not sure which is right but we should be consistent :)

File: docs/topics/api/addons.rst [hunk 5]
```diff
@@ -490,3 +490,35 @@ This endpoint returns a list of suggested replacements for legacy add-ons that a
...
+    :>json string results[].name: A description entered by AMO admins to describe the override.
+    :>json array results[].version_ranges: An array of affected versions of the add-on.
+    :>json string results[].version_ranges[].addon_min_version: minimum version of the add-on to be disabled.
+    :>json string results[].version_ranges[].addon_max_version: maximum version of the add-on to be disabled.
+    :>json array string results[].version_ranges[].applications: An array of affected applications for this range of versions.
+    :>json string results[].version_ranges[].applications[].name: Application name (e.g. Firefox).
+    :>json int results[].version_ranges[].applications[].id: Application id on AMO.
+    :>json string results[].version_ranges[].applications[].min_version: minimum version of the application to be disabled in.
+    :>json string results[].version_ranges[].applications[].max_version: maximum version of the application to be disabled in.
+    :>json string results[].version_ranges[].applications[].guid: Application `guid <https://addons.mozilla.org/en-US/firefox/pages/appversions/>`_.
```
Comment (diox) [inline docs/topics/api/addons.rst]: I know it's not your fault, but wow, this is confusing. We pass an add-on `guid` but this property is an application `guid`...

File: src/olympia/addons/models.py [hunk 6]
```diff
@@ -2083,6 +2083,11 @@ def transformer(overrides):
     def compat_ranges(self):
         return list(self._compat_ranges.all())
 
+    @property
+    def version_ranges(self):
```
Comment (diox) [inline src/olympia/addons/models.py]: Do you need that? Would `source='collapsed_ranges'` have worked in the serializer?

File: src/olympia/addons/views.py [hunk 7]
```diff
@@ -822,3 +823,16 @@ class ReplacementAddonView(ListAPIView):
...
+
+
+class CompatOverrideView(ListAPIView):
+    queryset = CompatOverride.objects.all()
+    serializer_class = CompatOverrideSerializer
+
+    def filter_queryset(self, queryset):
+        # Use the same Filter we use for AddonSearchView for consistency.
+        guid_filter = AddonGuidQueryParam(self.request)
+        guids = guid_filter.get_value()
```
Comment (diox) [inline src/olympia/addons/views.py]: What should we do when no guid is passed? Return a 400? This should be tested.

File: docs/topics/api/addons.rst (hunk 5 above)
Comment (eviljeff) [reply docs/topics/api/addons.rst]: I thought have the add-on guid be `addon-guid` and the application guid be inside `applications` was the clearest solution though I'm open to alternatives.  The legacy API has `guid` for the add-on guid and `appID` for the application guid (in addition to `application_id` for the AMO constant id).

File: src/olympia/addons/views.py (hunk 7 above)
Comment (eviljeff) [reply src/olympia/addons/views.py]: Hmm.  Not sure - currently you'd just get no results, as you would if didn't specify the `guid` parameter.  But I guess throwing an error is clearer when the endpoint only has one function.

File: src/olympia/addons/models.py (hunk 6 above)
Comment (eviljeff) [reply src/olympia/addons/models.py]: oh, it does.  I thought source needed a property for some reason.

File: src/olympia/addons/views.py (hunk 7 above)
Comment (diox) [reply src/olympia/addons/views.py]: Avoids a useless db query with `guid=''` as well :)

File: docs/topics/api/addons.rst (hunk 5 above)
Comment (diox) [reply docs/topics/api/addons.rst]: Yeah, re-reading the output I think it's the best we can do. Carry on:)

File: src/olympia/addons/views.py (hunk 7 above)
Comment (eviljeff) [reply src/olympia/addons/views.py]: https://github.com/mozilla/addons-server/pull/7339/commits/39781e9698733b84f8239ad2ac6232f9c64679ae

**Pull request merged**